File: examples/03-java-threads/Example01.java
// =================================================================
//
// File: Example01.java
// Author: Pedro Perez
// Description: This file implements the addition of two vectors 
//				using Java's Threads.
//
// Copyright (c) 2023 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

public class Example01 extends Thread {
	private static final int SIZE = 100_000_000;
	private int a[], b[], c[], start, end;

	public Example01(int start, int end, int c[], int a[], int b[]) {
		this.start = start;
		this.end = end;
		this.a = a;
		this.b = b;
		this.c = c;
	}

	public void run() {
		for (int i = start; i < end; i++) {
			c[i] = a[i] + b[i];
		}
	}

	public static void main(String args[]) {
		int a[] = new int [SIZE];
		int b[] = new int [SIZE];
		int c[] = new int [SIZE];
		long startTime, stopTime;
		double elapsedTime = 0;
		int blockSize;
		Example01 threads[];

		Utils.fillArray(a);
		Utils.displayArray("a", a);
		Utils.fillArray(b);
		Utils.displayArray("b", b);

		blockSize = SIZE / Utils.MAXTHREADS;
		threads = new Example01[Utils.MAXTHREADS];

		elapsedTime = 0;
		System.out.printf("Starting...\n");
		for (int j = 0; j < Utils.N; j++) {
			startTime = System.currentTimeMillis();

			for (int i = 0; i < threads.length; i++) {
				if (i != threads.length - 1) {
					threads[i] = 
					new Example01((i * blockSize), ((i + 1) * blockSize), c, a, b);
				} else {
					threads[i] = new Example01((i * blockSize), SIZE, c, a, b);
				}
			}

			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();

			elapsedTime += (stopTime - startTime);
		}
		Utils.displayArray("c", c);
		System.out.printf("avg time = %.5f ms\n", (elapsedTime / Utils.N));
	}
}

File: examples/06-tbb/example01.cpp
// =================================================================
//
// File: example01.cpp
// Author: Pedro Perez
// Description: This file implements the addition of two vectors 
//				using the TBB technology. To compile:
//				g++ example01.cpp -o app -I/usr/local/lib/tbb/include -L/usr/local/lib/tbb/lib/intel64/gcc4.4 -ltbb
//
// Copyright (c) 2023 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <chrono>
#include <tbb/blocked_range.h>
#include <tbb/parallel_for.h>
#include "utils.h"

using namespace std;
using namespace std::chrono;
using namespace tbb;

#define SIZE 10000000 // 1e7

class AddVectors {
private:
	int *vectorR, *vectorA, * vectorB;

public:
	AddVectors(int *result, int *a, int *b) 
		: vectorR(result), vectorA(a), vectorB(b) {} 

	void operator() (const blocked_range<int> &r) const {
		for (int i = r.begin(); i != r.end(); i++) {
			vectorR[i] = vectorA[i] + vectorB[i];
		}
	}
};

int main(int argc, char* argv[]) {
	int *a, *b, *c;

	// These variables are used to keep track of the execution time.
	high_resolution_clock::time_point start, end;
	double timeElapsed;

	a = new int [SIZE];
	b = new int [SIZE];
	c = new int [SIZE];

	fill_array(a, SIZE);
	display_array("a:", a);
	fill_array(b, SIZE);
	display_array("b:", b);

	cout << "Starting...\n";
	timeElapsed = 0;
	for (int j = 0; j < N; j++) {
		start = high_resolution_clock::now();

		parallel_for(blocked_range<int>(0, SIZE),  AddVectors(c, a, b));

		end = high_resolution_clock::now();
		timeElapsed += 
			duration<double, std::milli>(end - start).count();
	}
	display_array("c:", c);
	cout << "avg time = " << fixed << setprecision(3) 
		 << (timeElapsed / N) <<  " ms\n";

	delete [] a;
	delete [] b;
	delete [] c;

	return 0;
}
